hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/TestMixedQueueResourceCalculation.java

```diff
@@ -0,0 +1,424 @@
+        .assertEffectiveMinResource()
+        .toExpect(createResource(MEMORY * 0.5, VCORES))
+        .assertEffectiveMaxResource()
+        .build();
+
+    QueueCapacityUpdateContext updateContext = update(assertionBuilder, UPDATE_RESOURCE);
+    Optional<QueueUpdateWarning> queueA11ExceedsParentMaxResourceWarning = getSpecificWarning(
+        updateContext.getUpdateWarnings(), QueueUpdateWarningType.QUEUE_MAX_RESOURCE_EXCEEDS_PARENT,
+        A11);
+    Optional<QueueUpdateWarning> queueA11MinExceedsMaxWarning = getSpecificWarning(
+        updateContext.getUpdateWarnings(), QueueUpdateWarningType.QUEUE_EXCEEDS_MAX_RESOURCE, A11);
+    Assert.assertTrue(queueA11ExceedsParentMaxResourceWarning.isPresent());
+    Assert.assertTrue(queueA11MinExceedsMaxWarning.isPresent());
+  }
+
+  @Test
+  public void testNodeLabels() throws Exception {
+    setLabeledQueueConfigs();
+
+    QueueAssertionBuilder assertionBuilder = createAssertionBuilder()
+        .withQueue(A)
+        .toExpect(createResource(2048, 8))
+        .assertEffectiveMinResource(NO_LABEL)
+        .withQueue(A1)
+        .toExpect(createResource(1024, 5))
+        .assertEffectiveMinResource(NO_LABEL)
+        .withQueue(A2)
+        .toExpect(createResource(1024, 2))
+        .assertEffectiveMinResource(NO_LABEL)
+        .withQueue(B)
+        .toExpect(createResource(3072, 8))
+        .assertEffectiveMinResource(NO_LABEL)
+        .withQueue(A)
+        .toExpect(createResource(30720, 30))
+        .assertEffectiveMinResource(X_LABEL)
+        .withQueue(A1)
+        .toExpect(createResource(20480, 0))
+        .assertEffectiveMinResource(X_LABEL)
+        .withQueue(A2)
+        .toExpect(createResource(10240, 30))
+        .assertEffectiveMinResource(X_LABEL)
+        .withQueue(B)
+        .toExpect(createResource(30720, 30))
+        .assertEffectiveMinResource(X_LABEL)
+        .withQueue(A)
+        .toExpect(createResource(8096, 42))
+        .assertEffectiveMinResource(Y_LABEL)
+        .withQueue(A1)
+        .toExpect(createResource(6186, 21))
+        .assertEffectiveMinResource(Y_LABEL)
+        .withQueue(A2)
+        .toExpect(createResource(1910, 21))
+        .assertEffectiveMinResource(Y_LABEL)
+        .withQueue(B)
+        .toExpect(createResource(12384, 18))
+        .assertEffectiveMinResource(Y_LABEL)
+        .withQueue(A)
+        .toExpect(createResource(7168, 11))
+        .assertEffectiveMinResource(Z_LABEL)
+        .withQueue(A1)
+        .toExpect(createResource(6451, 4))
+        .assertEffectiveMinResource(Z_LABEL)
+        .withQueue(A2)
+        .toExpect(createResource(716, 7))
+        .assertEffectiveMinResource(Z_LABEL)
+        .withQueue(B)
+        .toExpect(createResource(3072, 4))
+        .assertEffectiveMinResource(Z_LABEL)
+        .build();
+
+    update(assertionBuilder, UPDATE_RESOURCE, Resource.newInstance(5 * GB, 16));
+  }
+
+  private void setLabeledQueueConfigs() throws Exception {
```

Comment:
I think creating a comment with the queue structure would make this test easier to read, like in TestCapacitySchedulerNewQueueAutoCreation